modules/cas_cache/layer_cache_management.c

```diff
@@ -15,141 +15,378 @@ extern u32 seq_cut_off_mb;
+	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
 
-	init_completion(&context.compl);
-	context.result = &result;
+	context = kmalloc(sizeof(*context), GFP_KERNEL);
+	if (!context)
+		return -ENOMEM;
 
-	atomic_set(&cache_priv->flush_interrupt_enabled, 0);
-	ocf_mngt_cache_flush(cache, _cache_mngt_cache_flush_complete, &context);
-	wait_for_completion_interruptible(&context.compl);
-	atomic_set(&cache_priv->flush_interrupt_enabled, 1);
+	init_completion(&context->compl);
+	atomic_set(&context->ref_count, 2);
+	atomic_set(&cache_priv->flush_interrupt_enabled, interruption);
+
+	ocf_mngt_cache_flush(cache, _cache_mngt_cache_flush_complete, context);
+	result = wait_for_completion_interruptible(&context->compl);
+
+	if (result < 0)
+		result = -KCAS_ERR_WAITING_INTERRUPTED;
+	else
+		result = context->result;
+
+	if (atomic_dec_and_test(&context->ref_count))
+		kfree(context);
 
 	return result;
 }
 
 static void _cache_mngt_core_flush_complete(ocf_core_t core, void *priv,
 		int error)
 {
-	struct _cache_mngt_sync_context *context = priv;
+	struct _cache_mngt_async_context *context = priv;
 
-	*context->result = error;
-	complete(&context->compl);
+	if (atomic_dec_and_test(&context->ref_count)) {
+		kfree(context);
+	} else {
+		context->result = error;
+		complete(&context->compl);
+	}
 }
 
 static int _cache_mngt_core_flush_sync(ocf_core_t core, bool interruption)
 {
+	int result;
+	struct _cache_mngt_async_context *context;
 	ocf_cache_t cache = ocf_core_get_cache(core);
 	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
-	struct _cache_mngt_sync_context context;
-	int result;
 
-	init_completion(&context.compl);
-	context.result = &result;
+	context = kmalloc(sizeof(*context), GFP_KERNEL);
+	if (!context)
+		return -ENOMEM;
+
+	init_completion(&context->compl);
+	atomic_set(&context->ref_count, 2);
+	atomic_set(&cache_priv->flush_interrupt_enabled, interruption);
+
+	ocf_mngt_core_flush(core, _cache_mngt_core_flush_complete, context);
+	result = wait_for_completion_interruptible(&context->compl);
+
+	if (result < 0)
+		result = -KCAS_ERR_WAITING_INTERRUPTED;
+	else
+		result = context->result;
+
+	if (atomic_dec_and_test(&context->ref_count))
+		kfree(context);
 
-	atomic_set(&cache_priv->flush_interrupt_enabled, 0);
-	ocf_mngt_core_flush(core, _cache_mngt_core_flush_complete, &context);
-	wait_for_completion_interruptible(&context.compl);
 	atomic_set(&cache_priv->flush_interrupt_enabled, 1);
 
 	return result;
 }
 
+static void _cache_mngt_cache_priv_deinit(ocf_cache_t cache)
+{
+	struct cache_priv *cache_priv = ocf_cache_get_priv(cache);
+
+	vfree(cache_priv);
+}
+
+static int exit_instance_finish(ocf_cache_t cache, int error)
+{
+	struct cache_priv *cache_priv;
+	ocf_queue_t mngt_queue;
+	bool flush_status;
+
+	flush_status = ocf_mngt_cache_is_dirty(cache);
+	cache_priv = ocf_cache_get_priv(cache);
+	mngt_queue = cache_priv->mngt_queue;
+
+	if (error && error != -OCF_ERR_WRITE_CACHE)
+		BUG_ON(error);
```

Comment:
If https://github.com/Open-CAS/ocf/pull/332 will be merged, the only error value returned by  ocf_mngt_cache_stop() will be -OCF_ERR_WRITE_CACHE. Since we don't expect any other value here, I believe BUG_ON() is ok here.